ENH: Added personal access token support to GitHub Collector

waldbauer-certat opened the pull request:
As github basic authentication has been marked as deprecated by
GitHub, we now implemented the Personal Access Token authentication
method.

Fixes #1549

sebix commented on CHANGELOG.md:
```diff
@@ -39,6 +39,7 @@ CHANGELOG
 
 #### Collectors
 - `intelmq.bots.collectors.mail._lib`: Add support for unverified SSL/STARTTLS connections (PR#2055 by Sebastian Wagner).
+- `intelmq.bots.collectors.github_api._collector_github_api`: Added personal access token support (PR#2145 by Sebastian Waldbauer).
```
```suggestion
- `intelmq.bots.collectors.github_api._collector_github_api`: Added personal access token support (PR#2145 by Sebastian Waldbauer, fixes #1549).
```

sebix commented on docs/user/bots.rst:
```diff
@@ -378,8 +378,9 @@ Github API
-* `basic_auth_password:` GitHub account password (optional)
+* `basic_auth_username:` GitHub account username (optional) (deprecated)
+* `basic_auth_password:` GitHub account password (optional) (deprecated)
+* `personal_access_token:` GitHub account personal access token (optional)
```
```suggestion
* `personal_access_token:` GitHub account personal access token ([GitHub documentation: Creating a personal access token](https://developer.github.com/changes/2020-02-14-deprecating-password-auth/#removal))
```

sebix reviewed (changes requested):
Please add a note to the NEWS file on the parameter change

https://developer.github.com/changes/2020-02-14-deprecating-password-auth/#removal says:

> All password authentication will return a status code of 401 starting:
> 
>     November 13, 2020 at 16:00 UTC

So I guess the old auth does no longer work? If that's the case, please remove the old method, it's of no use.

the `feeds.yml` needs an update as well: https://intelmq.readthedocs.io/en/latest/user/feeds.html#id123

waldbauer-certat commented:
> Please add a note to the NEWS file on the parameter change
> 
> https://developer.github.com/changes/2020-02-14-deprecating-password-auth/#removal says:
> 
> > All password authentication will return a status code of 401 starting:
> > ```
> > November 13, 2020 at 16:00 UTC
> > ```
> 
> So I guess the old auth does no longer work? If that's the case, please remove the old method, it's of no use.
> 
> the `feeds.yml` needs an update as well: https://intelmq.readthedocs.io/en/latest/user/feeds.html#id123

Feeds will be auto build, no? Because its gitignored https://github.com/certtools/intelmq/blob/develop/.gitignore#L30

sebix commented:
> > So I guess the old auth does no longer work? If that's the case, please remove the old method, it's of no use.
> > the `feeds.yml` needs an update as well: https://intelmq.readthedocs.io/en/latest/user/feeds.html#id123
> 
> Feeds will be auto build, no? Because its gitignored https://github.com/certtools/intelmq/blob/develop/.gitignore#L30

That page is generated from `intelmq/etc/feeds.yml`

sebix commented on intelmq/tests/bots/collectors/github_api/test_collector.py:
```diff
@@ -124,15 +112,6 @@ def test_collector_should_fail_with_bad_repository_error(self, requests_get_mock
         self.run_bot(parameters=SHOULD_FAIL_BECAUSE_REPOSITORY_IS_NOT_VALID_CONFIG['CONFIG'], prepare=True)
         self.assertRegexpMatchesLog(pattern=".*Unknown repository.*")  # assert the expected ValueError msg
 
-    @patch('intelmq.bots.collectors.github_api.collector_github_contents_api.requests.get')
-    def test_collector_should_fail_with_bad_credentials(self, requests_get_mock):
-        requests_get_mock.return_value.json = MagicMock(return_value={'message': 'Bad Credentials'})
-        requests_get_mock.return_value.configure_mock(status_code=401)
-
-        self.allowed_error_count = 1
-        self.run_bot(parameters=SHOULD_FAIL_WITH_BAD_CREDENTIALS['CONFIG'], prepare=True)
-        self.assertRegexpMatchesLog(pattern=".*Bad Credentials.*")
-
```
Is there no replacement possible?

sebix commented on intelmq/etc/feeds.yaml:
```diff
@@ -723,21 +723,21 @@ providers:
         contains a list of tags for each address which
         indicate what behaviour of the address was observed.
-        
+
```
are you sure that these whitespace removals do not affect the rendering?

waldbauer-certat committed:
ENH: Added personal access token support to GitHub Collector

As github basic authentication has been marked as deprecated by
GitHub, we now implemented the Personal Access Token authentication
method.

Fixes #1549

Signed-off-by: Sebastian Waldbauer <waldbauer@cert.at>

sebix commented on CHANGELOG.md:
```diff
@@ -42,6 +42,8 @@ CHANGELOG
   - Add support for unverified SSL/STARTTLS connections (PR#2055 by Sebastian Wagner).
   - Fix exception handling for aborted IMAP connections (PR#2187 by Sebastian Wagner).
 - `intelmq.bots.collectors.blueliv`: Fix Blueliv collector requirements (PR#2161 by Gethvi).
+- `intelmq.bots.collectors.mail._lib`: Add support for unverified SSL/STARTTLS connections (PR#2055 by Sebastian Wagner).
```
I don't see this change in this PR

sebix commented on intelmq/bots/collectors/github_api/_collector_github_api.py:
```diff
@@ -1,13 +1,12 @@
 """
-import base64
-
+from intelmq.lib.exceptions import MissingDependencyError
```
```suggestion
```

sebix commented on intelmq/bots/collectors/github_api/_collector_github_api.py:
```diff
@@ -23,16 +22,15 @@
 
     def init(self):
         if requests is None:
-            raise ValueError('Could not import requests. Please install it.')
+            raise MissingDependencyError("requests")
 
```
requests is a dependency of IntelMQ Core itself already

sebix commented on intelmq/bots/collectors/github_api/_collector_github_api.py:
```diff
@@ -23,16 +22,15 @@
         self.__user_headers = static_params['headers']
-        if self.basic_auth_username is not None and self.basic_auth_password is not None:
-            self.__user_headers.update(self.__produce_auth_header(self.basic_auth_username, self.basic_auth_password))
+        if self.personal_access_token is not None:
```
Treating empty string equally to None
```suggestion
        if self.personal_access_token:
```

sebix commented on intelmq/bots/collectors/github_api/_collector_github_api.py:
```diff
@@ -23,16 +22,15 @@
 class GithubAPICollectorBot(CollectorBot):
-    basic_auth_username = None
-    basic_auth_password = None
+    personal_access_token: str = None
```
None is not a string
```suggestion
    personal_access_token: Optional[str] = None
```

sebix commented:
And:
```
==================================== ERRORS ====================================
______ ERROR at setup of test_collector_should_fail_with_bad_credentials _______
file /home/runner/work/intelmq/intelmq/intelmq/tests/bots/collectors/github_api/test_collector.py, line 96
  @patch('intelmq.bots.collectors.github_api.collector_github_contents_api.requests.get')
  def test_collector_should_fail_with_bad_credentials(self, requests_get_mock):
E       fixture 'requests_get_mock' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, requests_mock, time_machine, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
/home/runner/work/intelmq/intelmq/intelmq/tests/bots/collectors/github_api/test_collector.py:96
---------- coverage: platform linux, python 3.7.13-final-0 -----------
Coverage XML written to file coverage.xml
=========================== short test summary info ============================
ERROR intelmq/tests/bots/collectors/github_api/test_collector.py::test_collector_should_fail_with_bad_credentials
```

sebix commented on intelmq/tests/bots/collectors/github_api/test_collector.py:
```diff
@@ -123,7 +119,7 @@ def test_collector_should_fail_with_bad_repository_error(self, requests_get_mock
         self.run_bot(parameters=SHOULD_FAIL_BECAUSE_REPOSITORY_IS_NOT_VALID_CONFIG['CONFIG'], prepare=True)
         self.assertRegexpMatchesLog(pattern=".*Unknown repository.*")  # assert the expected ValueError msg
-
+        
```
```suggestion

```

sebix commented on intelmq/tests/bots/collectors/github_api/test_collector.py:
```diff
@@ -95,7 +92,6 @@ def print_requests_get_parameters(url, *args, **kwargs):
         main_mock = MagicMock(content=EXAMPLE_CONTENT_STR)
         return main_mock
 
-
```
As [you are keen on codestyle and "Quality of Life"](https://github.com/certtools/intelmq/pull/2211): Before a class statement there should be two empty lines ;)
